validation/poststart_fail.go

```diff
@@ -0,0 +1,48 @@
+package main
+
+import (
+	"fmt"
+	"time"
+
+	tap "github.com/mndrix/tap-go"
+	rspec "github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/opencontainers/runtime-tools/specerror"
+	"github.com/opencontainers/runtime-tools/validation/util"
+	uuid "github.com/satori/go.uuid"
+)
+
+func main() {
+	t := tap.New()
+	t.Header(0)
+
+	config := util.LifecycleConfig{
+		Actions: util.LifecycleActionCreate | util.LifecycleActionStart | util.LifecycleActionDelete,
+		PreCreate: func(r *util.Runtime) error {
+			r.SetID(uuid.NewV4().String())
+			g := util.GetDefaultGenerator()
+			poststart := rspec.Hook{
+				Path: fmt.Sprintf("%s/%s/bin/false", r.BundleDir, g.Spec().Root.Path),
+				Args: []string{"false"},
+			}
+			g.AddPostStartHook(poststart)
+			g.SetProcessArgs([]string{"true"})
+			r.SetConfig(g)
+			return nil
+		},
+		PreDelete: func(r *util.Runtime) error {
+			util.WaitingForStatus(*r, util.LifecycleStatusStopped, time.Second*10, time.Second)
+			return nil
+		},
+	}
+
+	err := util.RuntimeLifecycleValidate(nil, config)
+	if err != nil {
+		err := specerror.NewError(specerror.PoststartHookFailGenWarn, fmt.Errorf("if any poststart hook fails, the runtime MUST log a warning, but the remaining hooks and lifecycle continue as if the hook had succeeded"), rspec.Version)
```

Comment:
Instead of assuming any error is the error we're trying to trigger, we should add a `PostCreate` action that flips a boolean:

```go
successfulCreate := false
config := util.LifecycleConfig{
  …
  PostCreate: func(r *util.Runtime) error {
    successfulCreate = true
    return nil
  },
}
…
if !successfulCreate {
  …fail the test…
} else if err != nil {
  …what you have here…
}
```

Reply:
We don't need to check this, we just need to verify if the 'poststart_fail', the remain hook and the lifecycle is working as expected. The sequence tests of 'create' and 'hooks' are covered in prestart.go/poststart.go/poststop.go .